tactics/declare.mli

```diff
@@ -81,6 +81,14 @@ val declare_private_constant
   -> unit proof_entry
   -> Constant.t * Evd.side_effects
 
+(** *)
+val inline_private_constants
+  :  sideff:bool
```

Comment:
This needs a big warning that it won't actually inline if you pass sideff:false

Reply:
Maybe the name of the function is wrong, it should rather be something like `handle_private_constants`.

Reply:
Ok will fix, on the other hand, is it even correct the current use with `sideff:false`?

Reply:
I'm unsure about the `handle_private_constants` naming tho, in the sense this doesn't really handle the constants but inline them, right? Handle seems to imply that they should always be inlined which I understand that is not the case?

Reply:
It may make more sense to only factorize the `sideff:true` part and leave the branch in build_by_tactic.
AFAICT build_by_tactic has some special behaviour which shouldn't be generally available.

Reply:
> It may make more sense to only factorize the `sideff:true` part and leave the branch in build_by_tactic.

Yes. But the caller with `false` seems very suspicious so I'm gonna research that before introducing back the if.

Reply:
cf https://github.com/coq/coq/commit/6459c0e8c240f0997873c50d4ec749effaba2f44

Reply:
> cf [6459c0e](https://github.com/coq/coq/commit/6459c0e8c240f0997873c50d4ec749effaba2f44)

Umm, not sure what the commit is trying to achieve, @ppedrot , does that make any sense? Why the schemes are generating side-effects? Are they calling abstract?

Reply:
Note that in the past very weird stuff was going on with schemes, and most of it was solved "piñata" style, see for example #9836 .

Reply:
Ok I updated the the PR with a warning, how does that look @SkySkimmer .

The story with the `sideff` parameter is kind tricky and goes back to its addition to `UState.merge` in 91f5467917266a85496fb718dfc30eff3565d4dc .

I think we should remove that flag, in most cases such as in `merge_context_set` it seems never set to true.

Reply:
But here is precisely the case where we use `sideff:true`

Reply:
Sorry I was confusing, there are two things:
- remove `sideff` from `build_by_tactic`, I plan to do this soon as the code in schemes seems suspicious [why is that different from regular transparent constant handling] This will require some work, as witnessed in abstract, the whole handling of opacity in that function is a bit messed up.
- investigate why duplicated universes are generated; that's a much more ambitious goal.

Reply:
>investigate why duplicated universes are generated; that's a much more ambitious goal.

cf https://github.com/coq/coq/wiki/Notes-on-side-effects,-universes-and-proof-closing btw

Reply:
Cool page, merci!